clients/src/main/java/org/apache/kafka/clients/consumer/StubbedAsyncKafkaConsumer.java

```diff
@@ -0,0 +1,548 @@
+        subscribe(pattern, new NoOpConsumerRebalanceListener());
+    }
+
+    /**
+     * Answers to the above questions:
+     *
+     * <ol>
+     *     <li>No</li>
+     *     <li>Yes, except in cases of bad input data</li>
+     *     <li>{@link Pattern} and {@link ConsumerRebalanceListener}</li>
+     *     <li><i>n/a</i></li>
+     * </ol>
+     */
+    @Override
+    public void subscribe(Pattern pattern, ConsumerRebalanceListener callback) {
+        maybeThrowInvalidGroupIdException();
+        if (pattern == null || pattern.toString().equals(""))
+            throw new IllegalArgumentException("Topic pattern to subscribe to cannot be " + (pattern == null ?
+                    "null" : "empty"));
+
+        throwIfNoAssignorsConfigured();
+        eventHandler.add(new SubscribePatternEvent(pattern, callback));
+    }
+
+    /**
+     * Answers to the above questions:
+     *
+     * <ol>
+     *     <li>No</li>
+     *     <li>Yes</li>
+     *     <li>None</li>
+     *     <li>None</li>
+     * </ol>
+     */
+    @Override
+    public void unsubscribe() {
+        eventHandler.add(new UnsubscribeEvent());
+    }
+
+    /**
+     * Answers to the above questions:
+     *
+     * <ol>
+     *     <li>No</li>
+     *     <li>Yes</li>
+     *     <li>List of {@link TopicPartition partitions}</li>
+     *     <li>None</li>
+     * </ol>
+     */
+    @Override
+    public void assign(Collection<TopicPartition> partitions) {
+        if (partitions == null) {
+            throw new IllegalArgumentException("Topic partition collection to assign to cannot be null");
+        } else if (partitions.isEmpty()) {
+            this.unsubscribe();
+        } else {
+            for (TopicPartition tp : partitions) {
+                String topic = (tp != null) ? tp.topic() : null;
+                if (Utils.isBlank(topic))
+                    throw new IllegalArgumentException("Topic partitions to assign to cannot have null or empty topic");
+            }
+
+            eventHandler.add(new AssignPartitionsEvent(partitions));
+        }
+    }
+
+    @Deprecated
+    @Override
+    public ConsumerRecords<K, V> poll(final long timeoutMs) {
+        return poll(time.timer(timeoutMs), false);
+    }
+
+    @Override
+    public ConsumerRecords<K, V> poll(Duration timeout) {
+        return poll(time.timer(timeout), true);
+    }
+
+    /**
+     * Answers to the above questions:
+     *
+     * <ol>
+     *     <li>Yes</li>
+     *     <li>Yes</li>
+     *     <li>List of {@link TopicPartition partitions}</li>
+     *     <li>None</li>
+     * </ol>
+     */
+    private ConsumerRecords<K, V> poll(final Timer timer, final boolean includeMetadataInTimeout) {
+        Map<TopicPartition, List<ConsumerRecord<K, V>>> records = new HashMap<>();
+
+        // The background thread will do the following when it receives this event:
+        //
+        // 1. Execute the 'update assignment metadata' logic (which I don't really understand at all, yet...)
```

Comment:
sorry, i think i'm not getting this part `for the former, the source of truth is on the background thread` (I assume you mean by subscription).  The current expectation is, as previously discussed, subscribe() -> subscriptions() should yield the most up to date subscription.  Changing the ownership will break this assumption no?

Reply:
Sorry I was not clear before, by `source of truth` I meant the assignment, i.e. list of partition infos, not the subscription, i.e. list of topics of topic-regex. When user called `subscribe()`, and then followed `subscription()` should definitely reflect the newly updated subscription info from `subscribe()`, but the followed `assignment()` would only be updated when the rebalance is completed, in which case it will be updated by the background thread.

Reply:
Got it, thanks for clarifying. I layout my understanding and expectation here: 

1. `assign()` -> `assignment()` : return the latest assigned partitions.
2. `subscribe()` -> `assignment()` : It would still be the old assignment because poll hasn't been called yet, so nothing would have been updated
3. `subscribed()` -> poll() [rebalance finished, callback triggered] -> `assignment()`: send back the latest assignment